### PR TITLE
add slider-3d gui control

### PR DESCRIPTION
**Pull request by binyammathewos**
Added a 3D slider GUI control

### CODE REVIEW

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:44]: Is there anything stopping a user from mixing up min/max, or making max < min in general?

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:58]: Personally I think the value should adapt to the new min/max value, instead of having the new min/max value adapt to the current value (Clamp the value to the new range, instead of expanding the range to accommodate the current value).

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:44]: So by default, there are 100 different values the slider can be set to. Do we want that many as a default?

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:153]: Is it possible to have the value be dynamically updated, instead of just on release? While keeping it set to specific values, of course. I've personally always gotten annoyed with scroll bars/other draggables that don't show me what I want until I release the button, meaning I can't browse options as easily.

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:43]: nit: can we declare these inline number values (0, 100, 50) as constants?

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:58]: should we allow setting minimum and maximum independent of one another? it seems like with this Math min/max check developers can end up in scenarios where the values don't get set correctly if they are trying to set a range (going from (0, 10) to (20, 30) for example could result in (10, 30) if setting min before max).

I feel like these min/max checks could unintentionally cause confusing value behaviors in their current form.

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:43]: how do we want to handle scenarios where a minimum is specified but no maximum, etc. Do we need to do things here to ensure that the minimum is less than the maximum or that the value falls within the minimum and maximum range?

Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:43]: Agreed, I will move those as constants. thanks

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:77]: nit: do we need separate comments for get vs set? Or will this comment auto resolve for both get and set in an ide?

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:110]: >     [](start = 70, length = 4)

nit: remove extra spaces

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:139]: >        [](start = 24, length = 7)

nit: there are a lot of extra spaces throughout. please remove.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:44]: The property setter won't allow that. But you are right the constructor is missing this validation. I will address that in the next pr. thanks

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:114]: >2 [](start = 36, length = 1)

should these arbitrary scale values be set through the options provided in the constructor? regardless, can we get any fallback values declared as nonexported constants at the top of the file/class?

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:136]: >  [](start = 26, length = 1)

nit: its unclear why some double spaces exist for this material constructor/material constructors below

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:58]: Agreed, I will fix this in the next pr. Thanks

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [reply gui/src/3D/controls/slider3D.ts:44]: I feel like I have seen value intervals added here as well for other slider controls. we probably want this so that we can support high precision as well as integer value scenarios

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:175]: nit: wrong spacing

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:128]: >_createBarMaterial [](start = 39, length = 18)

should we expose the ability to specify a material as an option?

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:44]: I took this from 2D slider control, which allows the slider be set to 100 values by default. We don't have to use 100 values. Please let me know your recommendation.  Thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:153]: Agreed. May be we can add this as an option for the end user to pass when creating the slider. By default we can dynamically update the value onDrag. Thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:58]: Agreed, I have a similar issues reported by rickfromwork , I will address this in my next pr. thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:43]: Agreed, We are missing validation here. we can throw exception for invalid ranges. I will address this in the next pr. thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:77]: I followed the other 3D controls style. But I  am okay to put the comment for both setter and getter. Thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:110]: Agreed, thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:139]: Agreed, will do that. Thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:136]: I will address that in my next pr. thanks

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:128]: Agreed, I will expose the materials as a public members. Thanks

File: gui/src/3D/controls/slider3D.ts
Comment (Popov72) [inline gui/src/3D/controls/slider3D.ts:44]: Is it expected that we can't set `maximum` to 0?

If 0 is a valid value for `maximum`, you can use:
```javascript
this._maximum = options?.maximum ?? 100;
```

For `minimum`, you can also use:
```javascript
this._minimum = options?.minimum ?? 0;
```

File: gui/src/3D/controls/slider3D.ts
Comment (rgerd) [inline gui/src/3D/controls/slider3D.ts:110]: nit: I'd make this a `TransformNode` since you don't actually need a mesh.

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [reply gui/src/3D/controls/slider3D.ts:77]: I was thinking about this too. I think having a single comment is fine, since the documentation generated from this puts both the setter and getter into a single entry.

For example, Mesh.ts's isUnIndexed
https://doc.babylonjs.com/typedoc/classes/babylon.mesh#isunindexed

Comment (rickfromwork) [reply gui/src/3D/controls/slider3D.ts:77]: It does look like the comment gets duplicated, though...

File: gui/src/3D/controls/slider3D.ts
Comment (msDestiny14) [inline gui/src/3D/controls/slider3D.ts:143]: super tiny nit: spacing

**Comment by ryantrem** [comment]
Do you have any screenshots or animated gifs you can share here so we can see what it actually looks like?

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:24]: Agreed, thanks

**Comment by msDestiny14** [comment]
> Do you have any screenshots or animated gifs you can share here so we can see what it actually looks like?

Second on this.

File: gui/src/3D/controls/index.ts
Comment (rickfromwork) [inline gui/src/3D/controls/index.ts:17]: Please place alphabetically in the list

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts]: This issue has been resolved. Take a look at the updated version.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts]: This issue has been resolved. Take a look at the updated version.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:110]: Agreed with Rick, having a single for both setter and getter make sense.

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:78]: Ah, so the SLIDER_* values are used as limiting values, not just the initial state. Interesting. I want to say there should also be a check in here that the new minimum is not larger than the current maximum, but I remember talking already about how that check doesn't exist for the 2D slider...

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:128]: Would it be possible to re-orient the thumb/bar such that the axis are aligned, letting us write something more readable like "this._sliderBar.position.x - this._sliderBar.scaling.x / 2"?

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:221]: Why is the decimal rounding done to two decimal places?

File: gui/src/3D/controls/slider3D.ts
Comment (rickfromwork) [inline gui/src/3D/controls/slider3D.ts:230]: nit: If it's like this elsewhere that's fine, but I'm always a fan of condensing one-line conditionals:
this._sliderBar?.dispose();

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:78]: We are assuming the slider values are only positive numbers. Since the default min value is 0, we are using it here to limit the min value the user trying to set to the default min value. Since 2D slider don't prevent the user from setting min > max, I am assuming we can leave it like this. The slider won't crash but will not work as intended. I was trying to avoid this circular validation for min & max. It is very confusing for the end user.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:48]: Yes, I do agree with that. It was Rick's idea to pass this as optional parameters. I will reach out to him to see if he has strong argument against it.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:128]: Yes, we can. I even found it very confusing every time I see this line.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:221]: No specific reason. I looked up the MRTK slider example, which uses two decimal places for values. One idea I have is to let the use set this as an option. We can default it to two decimals and we don't even have to check if  we need rounding or not (I am using the min, max or step values to determine that)

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:230]: Nice, I like that.

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:196]: >new Vector3(1, 0, 0) [](start = 72, length = 20)

nit: fwiw, some of these more common constants have static values. Like this vector3 here exists as a static Vector3.Right() value.

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [reply gui/src/3D/controls/slider3D.ts:110]: sounds good

File: gui/src/3D/controls/slider3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/slider3D.ts:164]: >0.03 [](start = 91, length = 4)

Do we think that an external developer will need to customize these values? or can they be reset after slider control creation?

Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:164]: I was assuming the developer can use scaling get/set property to scale the control  up/down to the desired size. I think of these values as scale of 1 for the control as a whole. I don't think we need these to be customizable.  Let me know your thought.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:196]: Nice, I did not know there is values for this. I only know Vector3.Zero().

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts]: This issue might be resolved. Can you take a look at the update on this PR? Thanks,

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts]: This issue might be resolved. Can you take a look at the update on this PR? Thanks,

**Review by chrisfromwork — APPROVED** [review]
:shipit:

**Comment by binyammathewos** [comment]
Please find attached here the Slider3D control in action. Thanks

slider3d

File: gui/src/3D/controls/slider3D.ts
Comment (Popov72) [reply gui/src/3D/controls/slider3D.ts]: It is solved indeed as the constructor does not take an `options` object anymore :)

File: gui/src/3D/controls/slider3D.ts
Comment (RaananW) [inline gui/src/3D/controls/slider3D.ts:44]: Would it be possible to pass those values in the constructor, defaulting to those values? In the current implementation it will take 5 different calls if I want to change all values.

File: gui/src/3D/controls/slider3D.ts
Comment (RaananW) [reply gui/src/3D/controls/slider3D.ts:150]: I am for allowing customization in this case. Mainly because you are changing the scaling.y value. A user that will manipulate the scaling values as they see fit will have these values overwritten. Might be enough to define a `scaleFactor` that will be applied to all of those default values.

**Review by RaananW — COMMENTED** [review]
LGTM! with two comments waiting for your feedback

**Comment by sebavan** [comment]
LGTM as well, I ll merge a soon as @RaananW comments have been addressed.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:44]: To be consistent with the other GUI control, we decided not to pass these parameters in the constructor. The initial PR uses constructor parameters, but there was a code review feedback to remove it for the reason I mentioned earlier.

File: gui/src/3D/controls/slider3D.ts
Comment (binyammathewos) [reply gui/src/3D/controls/slider3D.ts:150]: The user can use getter/setter public property for scaling the control up or down.  Those default values are like a seed that will give the control the default look.

**Comment by RaananW** [comment]
I just noticed you are pushing the changes to the preview branch and not to master. Any reason for that?

**Comment by sebavan** [comment]
Such a good catch ranaan !!!!! yup please retarget the pr to master.

**Review by RaananW — CHANGES_REQUESTED** [review]
Sorry, just to make sure it is not merged to the wrong branch. You have my approval when merged to master

**Comment by binyammathewos** [comment]
> I just noticed you are pushing the changes to the preview branch and not to master. Any reason for that?

No, reason. I re-target the PR to master. Thanks

**Review by RaananW — APPROVED** [review]

**Comment by RaananW** [comment]
Just wanted to make sure all of the conversations are resolved before merging this. Are we good to go?

**Pull request closed without merging**
